### PR TITLE
Ins-42 class domain

### PR DESCRIPTION
**Pull request by LinnikD**
ClassDomain instance for storing new types of system with its proxy and factory

### CODE REVIEW

File: genesis/public/core/class.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+// newClassDomain creates new instance of ClassDomain
+func newClassDomain(parent object.Parent) (*classDomain, error) {
+	if parent == nil {
+		return nil, fmt.Errorf("parent must not be nil")
+	}
+
+	classDomain := &classDomain{
+		BaseDomain: *domain.NewBaseDomain(parent, ClassDomainName),
```
Comment (megge-dream) [inline genesis/public/core/class.go]: domain.NewBaseDomain func returns *BaseDomain, why second *?

File: genesis/public/core/class.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	}
+	return classDomain, nil
+}
+
+// GetClassID return string representation of ClassDomain's class.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: GetClassID return**s**

File: genesis/public/core/class.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &classDomainProxy{
+		instance: instance,
+	}, nil
+}
+
+// RegisterClass proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: RegisterClass is a ...

File: genesis/public/core/class.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+func (cdp *classDomainProxy) RegisterClass(fc factory.Factory) (string, error) {
+	return cdp.instance.RegisterClass(fc)
+}
+
+// GetClass proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: GetClass is a ...

File: genesis/public/core/class.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+func (cdp *classDomainProxy) GetClass(record string) (factory.Factory, error) {
+	return cdp.instance.GetClass(record)
+}
+
+// GetReference proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: GetReference is a ...

File: genesis/public/core/class.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+func (cdp *classDomainProxy) GetReference() *object.Reference {
+	return cdp.instance.GetReference()
+}
+
+// GetParent proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: GetParent is a ...

File: genesis/public/core/class.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+func (cdp *classDomainProxy) GetParent() object.Parent {
+	return cdp.instance.GetParent()
+}
+
+// GetClassID proxy call for instance method.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: GetClassID is a ...

File: genesis/public/core/class.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+		parent: parent,
+	}
+}
+
+func (cdf *classDomainFactory) GetParent() object.Parent {
+	// TODO: return real parent, fix tests
+	return nil
+}
+
+// GetClassID return string representation of ClassDomain's class.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: GetClassID return**s**

File: genesis/public/core/class.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+func (cdf *classDomainFactory) GetClassID() string {
+	return class.ClsDomainID
+}
+
+// GetReference returns nil for not published factory
+func (cdf *classDomainFactory) GetReference() *object.Reference {
+	return nil
+}
+
+// Create factory method for new ClassDomain instances.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: Create is a ...

File: genesis/public/core/instance.go [hunk 10]
```diff
@@ -132,11 +132,20 @@ func (idp *instanceDomainProxy) GetClassID() string {
...
+		parent: parent,
+	}
+}
+
+func (idf *instanceDomainFactory) GetParent() object.Parent {
+	// TODO: return real parent, fix tests
+	return nil
 }
 
 // GetClassID return string representation of InstanceDomain's class.
```
Comment (megge-dream) [inline genesis/public/core/instance.go]: GetClassID return**s**

File: genesis/public/core/reference.go [hunk 11]
```diff
@@ -146,25 +146,34 @@ func (rdp *referenceDomainProxy) InitGlobalMap(globalInstanceMap *map[string]obj
...
 
 // NewReferenceDomainFactory creates new factory for ReferenceDomain.
-func NewReferenceDomainFactory() factory.Factory {
-	return &referenceDomainFactory{}
+func NewReferenceDomainFactory(parent object.Parent) factory.Factory {
+	return &referenceDomainFactory{
+		parent: parent,
+	}
+}
+
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: Comment about GetParent func

File: genesis/public/core/reference.go [hunk 12]
```diff
@@ -146,25 +146,34 @@ func (rdp *referenceDomainProxy) InitGlobalMap(globalInstanceMap *map[string]obj
...
 	return class.ReferenceDomainID
 }
 
 // GetReference returns nil for not published factory.
-func (adf *referenceDomainFactory) GetReference() *object.Reference {
+func (rdf *referenceDomainFactory) GetReference() *object.Reference {
 	return nil
 }
 
 // Create factory method for new ReferenceDomain instances.
```
Comment (megge-dream) [inline genesis/public/core/reference.go]: Create is a ...

Comment (LinnikD) [reply genesis/public/core/reference.go]: fixed

File: genesis/public/core/class.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+)
+
+// ClassDomainName is a name for class domain.
+const ClassDomainName = "ClassDomain"
+
+// ClassDomain is a contract that allows to publish new classes (e.g. new contract types).
+type ClassDomain interface {
+	// Base domain implementation.
+	domain.Domain
+	// RegisterClass is used to publish new .
```
Comment (megge-dream) [inline genesis/public/core/class.go]: new what?

File: genesis/public/core/class.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	return result, nil
+}
+
+type classDomainProxy struct {
+	instance *classDomain
+}
+
+// newClassDomainProxy creates new proxy and associate it with new instance of ClassDomain.
```
Comment (megge-dream) [inline genesis/public/core/class.go]: and associate**s**

File: genesis/public/core/doc.go [hunk 15]
```diff
@@ -21,17 +21,29 @@ ReferenceDomain - system domain that allow global reference resolving.
...
 
+
+ClassDomain - system domain that stores base system types named Classes.
+
+Usage:
+
+	factory := NewClassDomainFactory(parent)
+	clsDomain := factory.Create(parentDomain)
+
+	record, err := clsDomain.RegisterReference(someReference)
```
Comment (megge-dream) [inline genesis/public/core/doc.go]: RegisterReference?

**Pull request merged**